pkg/handlers/v1/fetchbyip.go

```diff
@@ -0,0 +1,78 @@
+package v1
+
+import (
+	"context"
+	"net"
+
+	"github.com/asecurityteam/ipam-facade/pkg/domain"
+	"github.com/asecurityteam/ipam-facade/pkg/logs"
+)
+
+// IPAddressQuery contains an IP address on which to search for physical assets.
+type IPAddressQuery struct {
+	IPAddress string `json:"ipAddress"`
+}
+
+// PhysicalAssetDetails provides the response structure for PhysicalAsset records returned from storage.
+type PhysicalAssetDetails struct {
+	IP            string `json:"ip"`
+	ResourceOwner string `json:"resourceOwner"`
+	BusinessUnit  string `json:"businessUnit"`
+	Tags          tags   `json:"tags"`
+}
+
+// tags is the key-value pair structure that provides less important information than the
+// root keys of the PhysicalAssetDetails response.
+type tags struct {
+	SubnetID string `json:"subnetID"`
+	Network  string `json:"network"`
+	DeviceID string `json:"deviceID"`
+	Location string `json:"location"`
+}
+
+// FetchByIPAddressHandler uses its PhysicalAssetFetcher implementation to serve fetch requests for
+// physical assets by IP address.
+type FetchByIPAddressHandler struct {
+	PhysicalAssetFetcher domain.PhysicalAssetFetcher
+	LogFn                domain.LogFn
+}
+
+// Handle processes an incoming IPAddressQuery and returns a PhysicalAssetDetails response or an error.
+func (h *FetchByIPAddressHandler) Handle(ctx context.Context, query IPAddressQuery) (PhysicalAssetDetails, error) {
+	logger := h.LogFn(ctx)
+
+	if ip := net.ParseIP(query.IPAddress); ip == nil {
+		err := domain.InvalidInput{IP: query.IPAddress}
+		logger.Error(logs.InvalidInput{Reason: err.Error()})
+		return PhysicalAssetDetails{}, err
+	}
+
+	asset, err := h.PhysicalAssetFetcher.FetchPhysicalAsset(ctx, query.IPAddress)
+	switch err.(type) {
+	case nil:
+		response := physicalAssetToResponse(asset)
+		return response, nil
+	case domain.AssetNotFound:
+		logger.Error(logs.AssetNotFound{Reason: err.Error()})
```

Comment:
suggestion: since this results in a 4XX, I think an Info level log would help keep the log noise down.

Reply:
This 404 is a critical error that should be surfaced to humans immediately. This is the system of last resort, so if an IP address has no match here, that indicates that we have scanned something that we have no record of, and something worthy of investigation has occurred.

With that context, do you still suggest an Info log level?

Reply:
That makes sense.

> that indicates that we have scanned something that we have no record of

That sounds to me that the error is specific to the context of scanning. If in the future there are other consumers who trigger 404s, would an error still be appropriate in this case?

Reply:
You're absolutely right. This service being the last resort is out-of-band information/an implementation detail subject to change.

Surfacing the error in this service's consumer is a much better place. 👍